feature: Add new module proxmox_cluster for creating/joining PVE clusters.

gyptazy opened the pull request:
##### SUMMARY
feature: Add new module proxmox_cluster for creating/joining PVE clusters.

Fixes: #5

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
This adds a new module for creating and joining PVE clusters.

##### ADDITIONAL INFORMATION
With this module, you can now simply create a PVE cluster by running a single task without doing crazy stunts in the roles. The same applies to joining a cluster with one or more nodes.

```
- name: Create a Proxmox VE Cluster
  community.proxmox.proxmox_cluster:
    state: present
    api_host: "{{ primary_node }}"
    api_user: root@pam
    api_password: "{{ proxmox_password }}"
    api_ssl_verify: false
    link0: 10.10.1.1
    link1: 10.10.2.1
    cluster_name: "devcluster"

- name: Get join information of the Proxmox VE Cluster
  community.proxmox.proxmox_cluster_info:
    api_host: "{{ primary_node }}"
    api_user: root@pam
    api_password: "{{ proxmox_password }}"
  register: cluster_info

- name: Join a Proxmox VE Cluster
  community.proxmox.proxmox_cluster:
    state_present
    api_host: "{{ secondary_node }}"
    api_user: root@pam
    api_password: "{{ proxmox_password }}"
    master_ip: "{{ primary_node }}"
    fingerprint: "{{ cluster_fingerprint }}"
```

felixfontein commented on changelogs/fragments/6-add-proxmox-cluster-module.yml:
```diff
@@ -0,0 +1,3 @@
+---
+minor_changes:
+  - proxmox_cluster - Add new module proxmox_cluster for creating/joining PVE clusters (https://github.com/ansible-collections/community.proxmox/pull/6).
```
Generally new plugins/modules don't need a changelog fragment. It will be automatically created from the top-level `version_added` combined with the `short_description` field.

gyptazy replied:
Thanks, just saw that it failed when slurping the fragment in the test pipeline - removed it.

l00ptr commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+
+
+def create_cluster(proxmox, module):
+    cluster_name = module.params['cluster_name'] or module.params['api_host']
```
```suggestion
    cluster_name = module.params.get('cluster_name') or module.params.get('api_host')
```

gyptazy replied:
ack

l00ptr commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+def create_cluster(proxmox, module):
+    cluster_name = module.params['cluster_name'] or module.params['api_host']
+    payload = {'clustername': cluster_name}
+    link0 = module.params.get('link0')
+    link1 = module.params.get('link1')
+
+    if link0:
+        payload['link0'] = link0
+    if link1:
+        payload['link1'] = link1
```
may be:
```suggestion
    if (link0:=module.params.get('link0')) is not None:
        payload['link0'] = link0
    if (link1:=module.params.get('link1')) is not None:
        payload['link1'] = link1
```

gyptazy replied:
ack

l00ptr commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+    except Exception as e:
+        module.fail_json(msg="Error while creating cluster: {}".format(str(e)))
+
+    module.exit_json(changed=True, msg="Cluster '{}' created.".format(cluster_name), cluster=cluster_name)
```
```suggestion
    module.exit_json(changed=True, msg=f"Cluster '{cluster_name}' created.")
```

gyptazy replied:
f-Strings failed in the pipeline when running the Py 2.7 tests (see previous commits before squashing).

l00ptr commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+    try:
+        proxmox.cluster.config.post(**payload)
+    except Exception as e:
+        module.fail_json(msg="Error while creating cluster: {}".format(str(e)))
```
```suggestion
        module.fail_json(msg=f"Error while creating cluster: {e}")
```

gyptazy replied:
f-Strings failed in the pipeline when running the Py 2.7 tests (see previous commits before squashing).

l00ptr commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+    try:
+        data = proxmox.cluster.config.join.get()
+    except Exception as e:
+        module.fail_json(msg="Error while obtaining cluster join information: {}".format(str(e)))
```
same here, we can probably use a f-string to make it easier to read.

gyptazy replied:
f-Strings failed in the pipeline when running the Py 2.7 tests (see previous commits before squashing).

felixfontein replied:
f-strings do not work with Python 2.7. (Right now the collection stlil supports Python 2.7.)

l00ptr commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+
+
+def join_cluster(proxmox, module):
+    master_ip = module.params['master_ip']
+    fingerprint = module.params['fingerprint']
+    api_password = module.params['api_password']
+    cluster_name = module.params.get('cluster_name')
```
could you use `.ge(...)` on all params ?

gyptazy replied:
ack, makes absolutely sense

felixfontein replied:
I'd probably use `[]` instead of `get()` since the values are always there (with value `None` if not set by the user and when they don't have a default), and you instantly notice that you made a typo (while `get()` simply returns `None` in that case), or later rename a parameter and forget one instance.

But in the end it's up for discussion to set a standard for the collection on how to do this (always use `get()`, always use `[]`, or let every module do whatever it wants, or even let every line in every module do whatever it wants).

l00ptr commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+    api_password = module.params['api_password']
+    cluster_name = module.params.get('cluster_name')
+
+    if not master_ip or not fingerprint:
```
```suggestion
    if  master_ip is None or fingerprint is None:
```
is more explicit.

gyptazy replied:
personal preferences, I can deal with it ;)

l00ptr commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+        )
+
+    except Exception as e:
+        module.fail_json(msg="Error while joining the cluster: {}".format(str(e)))
```
may be use a f-string here too.

gyptazy replied:
f-Strings failed in the pipeline when running the Py 2.7 tests (see previous commits before squashing).

l00ptr commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+    if __MISSING_MODULE__:
+        module.fail_json(msg="Required Python module is missing: proxmoxer")
+
+    action = module.params['action']
+    api_host = module.params['api_host']
+    api_user = module.params['api_user']
+    api_password = module.params['api_password']
+    api_ssl_verify = module.params['api_ssl_verify']
```
probably better to use `module.params.get(...)` here too

gyptazy replied:
ack, let's streamline it

l00ptr reviewed (changes requested):
Hi,

i have quickly reviewed that MR, i will make a deeper) review later today.

Thx for your proposal

gyptazy commented:
Hey @l00ptr,

> i have quickly reviewed that MR, i will make a deeper) review later today.
> Thx for your proposal

Thanks for your time and your review input. I squashed and pushed the suggested fixes - definitely makes sense to streamline the style. However, the f-strings failed the previous pipeline tests when it comes to Py 2.7 support (initially they were f-strings). Just let me know how to deal with it.

Appreciate your time and efforts!

Thanks,
gyptazy

gyptazy commented:
Good morning @l00ptr,

any news here? Happy to adjust the things to have it quickly included (will require it soon in customer projects and don't want to add a temporary different source for maintaining).

I'm also fine to immediately provide a draft PR fixing it to py3 syntax using walrus operator and f-strings to have it ready in place after the decisions in #8 have been taken - so it can immediately be merged and won't be forgotten.

Everyone, have a nice weekend and also a big thanks to @felixfontein for his efforts and help in Matrix!

Cheers,
gyptazy

gyptazy commented:
Hey @l00ptr,

> Your PR LGTM, thx for you contribution 👍
> So let's make the python 3 specific stuff into another PR!

Thanks, happy to hear. Who's going to merge it ;)?
After merging, I can provide the Py3 based adjustments as you recommended :)

Cheers
gyptazy

gyptazy commented:
Thanks for approving, can we merge this?

ping: @WaitingAtTheCrossroads @Thulium-Drake @l00ptr @Lithimlin

felixfontein commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+
+DOCUMENTATION = r"""
+module: proxmox_cluster
+version_added: 0.0.1
```
One thing that should be updated:
```suggestion
version_added: 0.1.0
```

gyptazy replied:
alright, squashed && pushed

Thulium-Drake commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+    if module.check_mode:
+        module.exit_json(changed=True, msg="Cluster '{}' would be created (check mode).".format(cluster_name), cluster=cluster_name)
+
+    try:
```
How does this handle when creating a cluster that already exists? Will it error out, or just tell the user it's all fine and return 'ok'?

Thulium-Drake commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+    if module.check_mode:
+        module.exit_json(changed=True, msg="Node would join the cluster (check mode).", cluster=cluster_name)
+
+    try:
```
Same here, will this error out when the node is already a member of _a_ cluster (AFAIK the docs mention a node should be re-installed, or at least thoroughly cleaned before joining a different cluster)?

Thulium-Drake commented on tests/unit/plugins/modules/test_proxmox_cluster.py:
```diff
@@ -0,0 +1,90 @@
+proxmoxer = pytest.importorskip("proxmoxer")
+mandatory_py_version = pytest.mark.skipif(
+    sys.version_info < (2, 7),
+    reason="The proxmoxer dependency requires python2.7 or higher",
```
Shouldn't this be 3.7, as discussed earlier in the issues?

felixfontein replied:
I'd completely remove this.

felixfontein replied:
I created #17 to clean up similar issues in other unit tests.

gyptazy replied:
will remove this as suggested by @felixfontein

Thulium-Drake reviewed (changes requested):
Thanks for the work you did! I added some notes

felixfontein commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,302 @@
+  action:
+    description:
+      - Possible module actions to perform by the Proxmox VE API.
+    choices: ['create_cluster', 'join_cluster', "get_cluster_info"]
```
Sorry I didn't notice this earlier, but: modules should not have **actions**, but **states**. Also, as @Thulium-Drake pointed out, `get_cluster_info` should be moved to an `_info` module. See https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html#designing-module-interfaces and https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html#scoping-your-module-s.

gyptazy commented:
Pushed the suggested fixes:
* Splitted into `proxmox_cluster` and `proxmox_cluster_info`
* Adjusted collection name
* Enhanced `check_mode`
* Removed Py 2.7 remarks from unit tests

Proxmox module part is tested but not yet the unit tests because I currently do not have access to my Ansible dev env. With a bit luck it passes but I think I need to adjust the tests tomorrow.

Thulium-Drake commented:
@gyptazy Thanks for the updates! It looks good to me. :rocket: 
@l00ptr Are you also happy with the PR in it's current state, if yes, it can be merged

gyptazy commented:
Good morning,

> @gyptazy Thanks for the updates! It looks good to me. 🚀 @l00ptr Are you also happy with the PR in it's current state, if yes, it can be merged

thanks, happy to hear that this fits to you. Hopefully this also fits to @l00ptr - would kindly ask you to merge it is this is also fine to you.

Thanks,
gyptazy

felixfontein commented on meta/runtime.yml:
```diff
@@ -10,6 +10,8 @@ action_groups:
     - proxmox_backup
     - proxmox_backup_info
+    - proxmox_cluster
+    - proxmox_cluster_info
```
(When checking why the sanity tests didn't report this - they should - I found a bug in antsibull-nox: https://github.com/ansible-community/antsibull-nox/pull/80, and a bug in the antsibull-nox config in this collection: https://github.com/ansible-collections/community.proxmox/pull/90...)

gyptazy replied:
Thanks for this. Indeed, I wasn't aware of it.

gyptazy commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,224 @@
+
+DOCUMENTATION = r"""
+module: proxmox_cluster
+version_added: 0.1.0
```
ack

gyptazy commented on plugins/modules/proxmox_cluster_info.py:
```diff
@@ -0,0 +1,182 @@
+
+DOCUMENTATION = r"""
+module: proxmox_cluster_info
+version_added: 0.1.0
```
ack

gyptazy commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,224 @@
+    description:
+      - The hostname of the Proxmox VE node to perform actions.
+    type: str
+    required: true
```
fixed

gyptazy commented on plugins/modules/proxmox_cluster.py:
```diff
@@ -0,0 +1,224 @@
+  description: A short message.
+  returned: always
+  type: str
+  sample: "Cluster devcluster created."
```
ack

felixfontein commented on meta/runtime.yml:
```diff
@@ -10,6 +10,8 @@ action_groups:
     - proxmox_backup
     - proxmox_backup_info
+    - proxmox_cluster
+    - proxmox_cluster_info
```
```suggestion
    - proxmox_cluster_join_info
```

gyptazy replied:
ouch... thanks

felixfontein commented on plugins/modules/proxmox_cluster_join_info.py:
```diff
@@ -0,0 +1,168 @@
+"""
+
+RETURN = r"""
+proxmox_cluster_join:
```
Why not use a shorter name, like `cluster` or `join` or `cluster_join`? This is the part the users *always* have to type, next to the registered variable's name.

gyptazy replied:
Fair enough, I just had a look at https://github.com/ansible-collections/community.proxmox/blob/main/plugins/modules/proxmox_storage_contents_info.py#L64 which uses `proxmox_storage_content`. I personally don't mind and we can change it. But imho, I like having the scope (`proxmox`) as a prefix or the whole module name which is unique. In the end, it's up to you. If you want to have it shorter, I'll change it.

felixfontein replied:
Why do you like having the scope here? This is not a top-level name, but something in a dictionary whose name is picked by the user.

If this would be a fact, you would be totally right. But this isn't a fact, but a key in a dictionary which only has other keys like `msg` and `changed`.

felixfontein reviewed (commented):
Just two small things (the first one is also pointed out by the sanity tests :) ), then it looks good from my side.

felixfontein commented on plugins/modules/proxmox_cluster_join_info.py:
```diff
@@ -0,0 +1,168 @@
+    api_password: "{{ password | default(omit) }}"
+    api_token_id: "{{ token_id | default(omit) }}"
+    api_token_secret: "{{ token_secret | default(omit) }}"
+  register: cluster_join
```
Why not keep the longer name here?
```suggestion
  register: proxmox_cluster_join
```

gyptazy replied:
ack

gyptazy commented:
Good evening everyone,

any news? Can we merge this (already 2 approvals) or anything missing?
Happy to adjust if something is missing.

Thanks,
gyptazy
